fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexTest.java

```diff
@@ -748,7 +771,223 @@ private void validateDocsInPartition(Index index, int partitionId, Tuple groupin
+        final int partitionHighWatermark = random.nextInt(15) + 2;
+        final int repartitionCount = random.nextInt(15) + 1;
+        LOGGER.info(KeyValueLogMessage.of("Running randomizedRepartitionTest",
+                "isGrouped", isGrouped,
+                "isSynthetic", isSynthetic,
+                "partitionHighWatermark", partitionHighWatermark,
+                "repartitionCount", repartitionCount));
+        Index index;
+        if (isGrouped) {
+            if (isSynthetic) {
+                index = getJoinedIndex(String.valueOf(partitionHighWatermark));
+                schemaSetup = context -> openRecordStore(context, metaDataBuilder -> metaDataBuilder.addIndex(joinedMetadataHook(metaDataBuilder), index));
+            } else {
+                index = complexPartitionedIndex(String.valueOf(partitionHighWatermark));
+                schemaSetup = context -> rebuildIndexMetaData(context, COMPLEX_DOC, index);
+            }
+        } else {
+            if (isSynthetic) {
+                index = getJoinedIndexNoGroup(String.valueOf(partitionHighWatermark));
+                schemaSetup = context -> openRecordStore(context, metaDataBuilder -> metaDataBuilder.addIndex(joinedMetadataHook(metaDataBuilder), index));
+            } else {
+                index = COMPLEX_PARTITIONED_NOGROUP;
+                schemaSetup = context -> rebuildIndexMetaData(context, COMPLEX_DOC, index);
+            }
+        }
+
+        final RecordLayerPropertyStorage contextProps = RecordLayerPropertyStorage.newBuilder()
+                .addProp(LuceneRecordContextProperties.LUCENE_REPARTITION_DOCUMENT_COUNT, repartitionCount)
+                .build();
+
+        // Generate random documents
+        Map<Integer, Map<Tuple, Long>> ids = new HashMap<>();
+        final int transactionCount = random.nextInt(15) + 1;
+        final long start = Instant.now().toEpochMilli();
+        Map<Integer, Set<Long>> allExistingTimestamps = new HashMap<>();
+        for (int i = 0; i < transactionCount; i++) {
+            final int docCount = random.nextInt(10) + 1;
+            try (FDBRecordContext context = openContext(contextProps)) {
+                schemaSetup.accept(context);
+                for (int j = 0; j < docCount; j++) {
+                    final int group = isGrouped ? random.nextInt(random.nextInt(10) + 1) : 0; // irrelevant if !isGrouped
+                    final int countInGroup = ids.computeIfAbsent(group, key -> new HashMap<>()).size();
+                    // we currently don't support multiple records with the same timestamp, specifically at the boundaries
+                    long timestamp = start + countInGroup + random.nextInt(20) - 5;
+                    final Set<Long> existingTimestamps = allExistingTimestamps.computeIfAbsent(group, key -> new HashSet<>());
+                    while (!existingTimestamps.add(timestamp)) {
+                        timestamp++;
+                    }
```

Comment:
Since we don't currently support having two records at the boundary with the same timestamp this makes that easier.
If we improve that functionality, we'll probably want to change this.